certora/specs/Pausable.spec

```diff
@@ -0,0 +1,93 @@
+*/
+rule pause(env e) {
+    require nonpayable(e);
+
+    bool pausedBefore = paused();
+
+    pause@withrevert(e);
+    bool success = !lastReverted;
+
+    bool pausedAfter = paused();
+    
+    // liveness
+    assert success => !pausedBefore, "unpause call must succeed if the contract was unpaused before";
+
+    // effect
+    assert success => (pausedBefore != pausedAfter && pausedAfter == true), "contract must be paused after a successful call";
+
+    // no side effect
+    assert !success => (pausedBefore == pausedAfter && pausedAfter == true), "contract must kept paused after a failed call";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: _unpause unpauses the contract                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule unpause(env e) {
+    require nonpayable(e);
+
+    bool pausedBefore = paused();
+
+    unpause@withrevert(e);
+    bool success = !lastReverted;
+
+    bool pausedAfter = paused();
+    
+    // liveness
+    assert success => pausedBefore, "pause call must succeed if the contract was paused before";
+
+    // effect
+    assert success => (pausedBefore != pausedAfter && pausedAfter == false), "contract must be unpaused after a successful call";
+
+    // no side effect
+    assert !success => (pausedBefore == pausedAfter && pausedAfter == false), "contract must kept unpaused after a failed call";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: only _pause can pause                                                                                        │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noPause(env e) {
+    method f;
+    calldataarg args;
+
+    require !paused();
+
+    bool pausedBefore = paused();
+    f(e, args);
+    bool pausedAfter = paused();
+
+    assert pausedBefore != pausedAfter => f.selector == pause().selector, "contract must only be paused by _pause()";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: only _unpause can unpause                                                                                    │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noUnpause(env e) {
+    method f;
+    calldataarg args;
+
+    require paused();
+
+    bool pausedBefore = paused();
+    f(e, args);
+    bool pausedAfter = paused();
+
+    assert pausedBefore != pausedAfter => f.selector == unpause().selector, "contract must only be unpaused by _unpause()";

```

Comment:
these two rules can be merged into something like this

```
rule noUnpause(env e) {
    bool pausedBefore = paused();
    method f; calldataarg args; f(e, args);
    bool pausedAfter = paused();

    assert pausedBefore != pausedAfter => (
        (!pausedAfter && f.selector == unpause().selector) ||
        (pausedAfter && f.selector == pause().selector)
    );

Reply:
Just merged them

https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4117/commits/66f18ffb66f38480cdfa443262c7f49e038d8abc